### PR TITLE
Experimental flag for controlling initializers as inputs | feat(torchlib)

### PR DESCRIPTION
**Pull request by justinchuby**
Create a flag `TORCHLIB_EXPERIMENTAL_INITIALIZERS_AS_INPUTS=1` to mark initializers as inputs to the model.

**This flag is experimental only for ONNX Runtime training and should not be assumed to exist in production.**

Reference: https://github.com/microsoft/onnx-converters-private/issues/182

Tested with the open-llama model with `transformers==4.31.0` with script https://gist.github.com/abock/2115d34d98df15a77516e8a2899b121c

image

### CODE REVIEW

File: onnxscript/function_libs/torch_lib/graph_building.py [hunk 1]
```diff
@@ -749,14 +750,17 @@ def to_model_proto(
 
         large_model = initializers_size > _LARGE_MODEL_SIZE_THRESHOLD
 
+        # TODO(justinchuby): If initializers is empty, do we still know which ones are initializers? Should we use the defer_weight_export argument?
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/graph_building.py:753]:
```suggestion
```

We do

**Review by BowenBao — COMMENTED** [review]
I feel like though this requirement is more naturally satisfied on exporter side by adding an option that allows adjusting the `include_initializer` flag that decides this, wdyt?

File: onnxscript/function_libs/torch_lib/_flags.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+"""Experimental flags.
+
+NOTE: These flags are experimental only. Any flag here can be removed at any
+time without notice.
+"""
+
+import os
+
+EXPERIMENTAL_INITIALIZERS_AS_INPUTS: bool = (
```
Comment (BowenBao) [inline onnxscript/function_libs/torch_lib/_flags.py:9]: nit: add a comment explaining the behavior?

File: onnxscript/function_libs/torch_lib/_flags.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
+"""Experimental flags.
+
+NOTE: These flags are experimental only. Any flag here can be removed at any
```
Comment (BowenBao) [inline onnxscript/function_libs/torch_lib/_flags.py:3]: Should we emit warning whenever flag uses under this file is detected?

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/_flags.py:3]: Good idea

Comment (BowenBao) [reply onnxscript/function_libs/torch_lib/_flags.py:3]: And raise error when this flag is removed but still used by user.

**Review by BowenBao — APPROVED** [review]

File: onnxscript/function_libs/torch_lib/_flags.py (hunk 2 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/_flags.py]: Done. Added as string. PTAL

File: onnxscript/function_libs/torch_lib/_flags.py (hunk 3 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/_flags.py:3]: Done

File: onnxscript/function_libs/torch_lib/_flags.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    deprecated: bool = False,
+) -> bool:
+    """Load a boolean flag from environment variable.
+
+    Args:
+        name: The name of the environment variable.
+        this_will: A string that describes what this flag will do.
+        deprecated: Whether this flag is deprecated.
+    """
+    state = os.getenv(name) == 1
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/_flags.py:26]:
```suggestion
    state = os.getenv(name) == "1"
```

**Review by BowenBao — APPROVED** [review]

**Comment by justinchuby** [comment]
Thanks for reviewing!

**Pull request merged**